keel-tagging-plugin/src/main/kotlin/com/netflix/spinnaker/keel/tagging/ResourceTagger.kt

```diff
@@ -0,0 +1,188 @@
+ * that we're managing.
+ */
+class ResourceTagger(
+  private val resourceRepository: ResourceRepository,
+  private val resourcePersister: ResourcePersister,
+  private val cloudDriverService: CloudDriverService,
+  private var removedTagRetentionHours: Long = 24,
+  private var clock: Clock
+) {
+  private val log = LoggerFactory.getLogger(javaClass)
+
+  private var accounts: Set<Credential> = emptySet()
+  private var accountsUpdateTimeS = 0L
+  private var accountsUpdateFrequencyS = TimeUnit.MINUTES.toSeconds(10)
+  private val transforms = mapOf(
+    "ec2" to "aws"
+  )
+
+  @EventListener(CreateEvent::class)
+  fun onCreateEvent(event: CreateEvent) {
+    if (event.resourceName.isNotTag()) {
+      log.debug("Persisting tag desired for resource {} because it exists now", event.resourceName.toString())
+      val spec = event.resourceName.generateKeelTagSpec()
+      persistTagState(spec)
+    }
+  }
+
+  @EventListener(DeleteEvent::class)
+  fun onDeleteEvent(event: DeleteEvent) {
+    if (event.resourceName.isNotTag()) {
+      log.debug("Persisting no tag desired for resource {} because it is no longer managed", event.resourceName.toString())
+      val spec = KeelTagSpec(
+        keelId = event.resourceName.toString(),
+        entityRef = event.resourceName.toEntityRef(),
+        tagState = TagNotDesired(startTime = clock.millis())
+      )
+      persistTagState(spec)
+    }
+  }
+
+  private fun persistTagState(spec: KeelTagSpec) {
+    val submitted = spec.toSubmittedResource()
+    val name = ResourceName(spec.generateTagNameFromKeelId())
+
+    if (tagExists(name)) {
+      resourcePersister.update(name, submitted)
+    } else {
+      resourcePersister.create(submitted)
+    }
+  }
+
+  private fun tagExists(tagResourceName: ResourceName): Boolean {
+    try {
+      resourceRepository.get(tagResourceName, KeelTagSpec::class.java)
+      return true
+    } catch (e: NoSuchResourceException) {
+      return false
+    }
+  }
+
+  @Scheduled(fixedDelayString = "\${keel.resource-tagger.tag-cleanup-frequency:P1D}")
+  fun removeTags() {
+    log.info("Cleaning up old keel tags")
+    resourceRepository.allResources { resourceHeader ->
+      if (resourceHeader.apiVersion == SPINNAKER_API_V1.subApi("tag")) {
+        val tagResource = resourceRepository.get(resourceHeader.name, KeelTagSpec::class.java)
+        if (tagResource.spec.tagState is TagNotDesired) {
+          val tagState = tagResource.spec.tagState as TagNotDesired
+          if (tagState.startTime < clock.millis() - Duration.ofHours(removedTagRetentionHours).toMillis()) {
+            resourcePersister.delete(resourceHeader.name)
+          }
+        }
+      }
+    }
+  }
+
+  @Scheduled(fixedDelayString = "\${keel.resource-tagger.account-sync-frequency:PT600S}")
+  private fun syncAccounts() {
+    val now = LocalDateTime.now().toEpochSecond(ZoneOffset.UTC)
+    if (now - accountsUpdateFrequencyS > accountsUpdateTimeS) {
```

Comment:
Is there any need to check this given the method is called on a schedule by Spring?

Reply:
How would I test that?

Reply:
It should just always sync the accounts when the method is called, no?